### ISSUE TITLE
new SSL implementation gives AttributeError when using an invalid key file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5449]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wulczer reported|
|-|-|
|Trac ID|trac#5449|
|Type|release blocker: regression|
|Created|2012-01-10 16:29:28Z|
|Branch|https://github.com/twisted/twisted/tree/early-context-error-5449|

The attached snippet tries to open an SSL connection using /etc/passwd as the key file.

In Twisted 11.0.0 this gave a traceback with an OpenSSL.SSL.Error, in Twisted 11.1.0 it also tracebacks with an AttributeError.

Attachments:

* [ssl.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/20d/20dfd961e7d5ce4e11e5c93ccf47d3070983487e/40e7d6dace4ac3468fa51e774a8e908809fcd256.py) (778 bytes) - added by wulczer on 2012-01-10 16:29:57Z - snippet demonstrating the change in behaviour

<details><summary>Searchable metadata</summary>

```
trac-id__5449 5449
type__release_blocker__regression release blocker: regression
reporter__wulczer wulczer
priority__normal normal
milestone__Twisted_12_0 Twisted-12.0
branch__branches_early_context_error_5449 branches/early-context-error-5449
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1326212968000000 1326212968000000
changetime__1327193304000000 1327193304000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Here's the output:

```
Unhandled Error
Traceback (most recent call last):
  File "/home/therve/Projects/Twisted/trunk/twisted/python/log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/python/log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/posixbase.py", line 591, in _doReadOrWrite
    why = selectable.doWrite()
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/tcp.py", line 419, in doConnect
    self._connectDone()
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/tcp.py", line 427, in _connectDone
    self.protocol.makeConnection(self)
  File "/home/therve/Projects/Twisted/trunk/twisted/protocols/tls.py", line 293, in makeConnection
    tlsContext = self.factory._contextFactory.getContext()
  File "ssl.py", line 11, in getContext
    ctx.use_certificate_file('/etc/passwd')
OpenSSL.SSL.Error: [('PEM routines', 'PEM_read_bio', 'no start line'), ('SSL routines', 'SSL_CTX_use_certificate_file', 'PEM lib')]
Unhandled Error
Traceback (most recent call last):
  File "ssl.py", line 38, in <module>
    reactor.run()
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/base.py", line 1169, in run
    self.mainLoop()
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/base.py", line 1181, in mainLoop
    self.doIteration(t)
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/pollreactor.py", line 167, in doPoll
    log.callWithLogger(selectable, _drdw, selectable, fd, event)
--- <exception caught here> ---
  File "/home/therve/Projects/Twisted/trunk/twisted/python/log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/python/log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/posixbase.py", line 599, in _doReadOrWrite
    self._disconnectSelectable(selectable, why, inRead)
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/posixbase.py", line 266, in _disconnectSelectable
    selectable.connectionLost(failure.Failure(why))
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/tcp.py", line 433, in connectionLost
    Connection.connectionLost(self, reason)
  File "/home/therve/Projects/Twisted/trunk/twisted/internet/tcp.py", line 277, in connectionLost
    protocol.connectionLost(reason)
  File "/home/therve/Projects/Twisted/trunk/twisted/protocols/tls.py", line 455, in connectionLost
    self._tlsConnection.bio_shutdown()
  File "/home/therve/Projects/Twisted/trunk/twisted/protocols/policies.py", line 112, in __getattr__
    return getattr(self.transport, name)
exceptions.AttributeError: 'NoneType' object has no attribute '_tlsConnection'
TIMEOUT
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [33454]) Branching to 'early-context-error-5449'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [33455]) Make IReactorSSL.connectSSL raise a synchronous exception if the context factory passed to it is broken

refs [#5449](https://github.com/twisted/twisted/issues/5449)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

The behavior in 11.0.0 is:

```
Unhandled Error
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/python/log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/python/log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/internet/tcp.py", line 674, in doConnect
    self._connectDone()
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/internet/ssl.py", line 149, in _connectDone
    self.startTLS(self.ctxFactory)
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/internet/tcp.py", line 571, in startTLS
    if Connection.startTLS(self, ctx, client):
  File "/home/exarkun/Projects/Twisted/branches/twisted-11.0.0/twisted/internet/tcp.py", line 402, in startTLS
    self.socket = SSL.Connection(ctx.getContext(), self.socket)
  File "ssl.py", line 11, in getContext
    ctx.use_certificate_file('/etc/passwd')
OpenSSL.SSL.Error: [('PEM routines', 'PEM_read_bio', 'no start line'), ('SSL routines', 'SSL_CTX_use_certificate_file', 'PEM lib')]
TIMEOUT
```

However, that sucks.  It's quite unclear in what state the TCP connection is left, and it's a very difficult error for an application to handle, since it's merely logged as unhandled, not passed to any application code.

I don't think this behavior was ever considered, it merely occurred by accident.  Better behaviors would be:

  1. Raise the exception from `connectSSL`
  1. Pass the exception to the factory's `clientConnectionFailed`.

The latter of these two is rather tempting, however the former is much simpler to implement and more closely mirrors the server-side behavior.  

Hence, the behavior in the branch is now:

```
$ python ssl.py 
Traceback (most recent call last):
  File "ssl.py", line 36, in <module>
    reactor.connectSSL('twistedmatrix.com', 443, Factory(), CtxFactory())
  File "/home/exarkun/Projects/Twisted/branches/early-context-error-5449/twisted/internet/posixbase.py", line 452, in connectSSL
    tlsFactory = tls.TLSMemoryBIOFactory(contextFactory, True, factory)
  File "/home/exarkun/Projects/Twisted/branches/early-context-error-5449/twisted/protocols/tls.py", line 589, in __init__
    contextFactory.getContext()
  File "ssl.py", line 11, in getContext
    ctx.use_certificate_file('/etc/passwd')
OpenSSL.SSL.Error: [('PEM routines', 'PEM_read_bio', 'no start line'), ('SSL routines', 'SSL_CTX_use_certificate_file', 'PEM lib')]
```

I expect it would be possible to return exactly to the 11.0.0 (and earlier - apparently back as far as 8.0.0, where I stopped checking) behavior.  I'm not sure that's desirable, though.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/early-context-error-5449)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun|
|-|-|

Throwing immediately seems reasonable, given all the relevant APIs are synchronous. And it's a more clever solution than I would've come up with.

Remaining issues:

   1. Equivalent test for startTLS. Looking at the code, I assume it will just work. Could be done in separate ticket.
   1. listenSSL presumably has the same issue. Doing this in separate ticket would also be fine.
   1. News file.

Given fixes/new tickets for above, feel free to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [33468]) Merge early-context-error-5449

Author: exarkun
Reviewer: itamar
Fixes: [#5449](https://github.com/twisted/twisted/issues/5449)

Change the users of application-supplied SSL context factories to invoke
the `getContext` method early and allow errors to propagate.  This improves
the behavior in the case of broken context factories which cannot produce
a context object by making the error happen closer to where the broken
object is given to Twisted.
